qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,299 @@
+        """Resolve a DNS hostname.
+
+        Resolves the given DNS hostname into an IP address, and returns it
+        in the dot-separated format as a string.
+
+        Args:
+            host: hostname to resolve.
+        """
+        ips = QHostInfo.fromName(host)
+        if ips.error() != QHostInfo.NoError or not ips.addresses():
+            err_f = "Failed to resolve host during PAC evaluation: {}"
+            log.network.info(err_f.format(host))
+            return QJSValue(QJSValue.NullValue)
+        else:
+            return ips.addresses()[0].toString()
+
+    @_js_slot()
+    def myIpAddress(self):
+        """Get host IP address.
+
+        Return the server IP address of the current machine, as a string in
+        the dot-separated integer format.
+        """
+        return QHostAddress(QHostAddress.LocalHost).toString()
+
+
+class PACResolver(object):
+
+    """Evaluate PAC script files and resolve proxies."""
+
+    @staticmethod
+    def _parse_proxy_host(host_str):
+        host, _colon, port_str = host_str.partition(':')
+        try:
+            port = int(port_str)
+        except ValueError:
+            raise ParseProxyError("Invalid port number")
+        return (host, port)
+
+    @staticmethod
+    def _parse_proxy_entry(proxy_str):
+        """Parse one proxy string entry, as described in PAC specification."""
+        config = [c for c in proxy_str.split(' ') if c]
+        if not config:
+            raise ParseProxyError("Empty proxy entry")
+        elif config[0] == "DIRECT":
+            if len(config) != 1:
+                raise ParseProxyError("Invalid number of parameters")
+            return QNetworkProxy(QNetworkProxy.NoProxy)
+        elif config[0] == "PROXY":
+            if len(config) != 2:
+                raise ParseProxyError("Invalid number of parameters")
+            host, port = PACResolver._parse_proxy_host(config[1])
+            return QNetworkProxy(QNetworkProxy.HttpProxy, host, port)
+        elif config[0] == "SOCKS":
+            if len(config) != 2:
+                raise ParseProxyError("Invalid number of parameters")
+            host, port = PACResolver._parse_proxy_host(config[1])
+            return QNetworkProxy(QNetworkProxy.Socks5Proxy, host, port)
+        else:
+            err = "Unknown proxy type: {}"
+            raise ParseProxyError(err.format(config[0]))
+
+    @staticmethod
+    def _parse_proxy_string(proxy_str):
+        proxies = proxy_str.split(';')
+        return [PACResolver._parse_proxy_entry(x) for x in proxies]
+
+    def __init__(self, pac_str):
+        """Create a PAC resolver.
+
+        Args:
+            pac_str: JavaScript code containing PAC resolver.
+        """
+        self._engine = QJSEngine()
+
+        self._ctx = _PACContext(self._engine)
+        self._engine.globalObject().setProperty("PAC",
+            self._engine.newQObject(self._ctx))
+        self._engine.evaluate(_PACContext.JS_DEFINITIONS, "pac_js_definitions")
```

Comment:
The [QJSEngine::evaluate](http://doc.qt.io/qt-5/qjsengine.html#evaluate) docs say otherwise: It returns the result of the evaluation (which in our case is probably `undefined` or `null`), or an error object if one happened (Qt doesn't throw exceptions)